include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp

```diff
@@ -185,86 +185,43 @@ namespace alpaka
-            };
-
-            struct CallbackSynchronizationData : public std::enable_shared_from_this<CallbackSynchronizationData>
-            {
-                std::mutex m_mutex;
-                std::condition_variable m_event;
-                CallbackState m_state = CallbackState::enqueued;
+            using QueueImpl = uniform_cuda_hip::detail::QueueUniformCudaHipRtImpl<TApi>;
+
+            struct HostFuncData
+            {
+                QueueImpl& q; // We don't need to keep the queue alive, because in it's dtor it will synchronize with
+                              // the CUDA/HIP stream and wait until all host functions and the CallbackThread are done.
+                              // It's actually an error ot copy the queue into the host function. Destroying it here
+                              // would call CUDA/HIP APIs from the host function. Passing it further to the Callback
+                              // thread, would make the Callback thread hold a task containing the queue with the
+                              // CallbackThread itself. Destroying the task if no other queue instance exists will
+                              // make the CallbackThread join itself and crash.
+                TTask t;
             };
 
             ALPAKA_FN_HOST static void uniformCudaHipRtHostFunc(void* arg)
             {
-                // explicitly copy the shared_ptr so that this method holds the state even when the executing thread
-                // has already finished.
-                auto const spCallbackSynchronizationData
-                    = reinterpret_cast<CallbackSynchronizationData*>(arg)->shared_from_this();
-
-                // Notify the executing thread.
-                {
-                    std::unique_lock<std::mutex> lock(spCallbackSynchronizationData->m_mutex);
-                    spCallbackSynchronizationData->m_state = CallbackState::notified;
-                }
-                spCallbackSynchronizationData->m_event.notify_one();
-
-                // Wait for the executing thread to finish the task if it has not already finished.
-                std::unique_lock<std::mutex> lock(spCallbackSynchronizationData->m_mutex);
-                if(spCallbackSynchronizationData->m_state != CallbackState::finished)
-                {
-                    spCallbackSynchronizationData->m_event.wait(
-                        lock,
-                        [&spCallbackSynchronizationData]()
-                        { return spCallbackSynchronizationData->m_state == CallbackState::finished; });
-                }
-            }
+                auto data = std::unique_ptr<HostFuncData>(reinterpret_cast<HostFuncData*>(arg));
+                auto& queue = data->q;
+                auto f = queue.m_callbackThread.submit(
+                    [data = std::move(data)]() mutable
+                    {
+                        data->t();
+                        data = nullptr; // destroy the task
```

Comment:
```suggestion
                        data.reset(); // destroy the task
```
looks nicer to me :-p

Reply:
Alright!